### PR TITLE
Cherry-pick from PaddlePaddle:feature/anakin-engine: Anakin subgraph support.

### PR DESCRIPTION
**Pull request by NHZlX**
This branch Align with branch PaddlePaddle:feature/anakin-engine except for: 

```
1. conv_fusion cudnn workspace problem,  feature/anakin-engine branch takes a hard code way. A better way should be considered.
2. add int32 data type for zerocopy.
3. should add ut for paddle-anakin.
```

```
cmake .. -DANAKIN_ROOT=/anakin_root/
```

The 3 points above will be added in this pr later.

### CODE REVIEW

**Review by Shixiaowei02 — DISMISSED** [review]

File: paddle/fluid/inference/anakin/convert/test_fc_op.cc [hunk 1]
```diff
@@ -22,26 +21,24 @@ namespace inference {
...
 
   // Prepare Op description
   framework::OpDesc desc;
   desc.SetType("mul");
   desc.SetInput("X", {"mul_x"});
   desc.SetInput("Y", {"mul_y"});
   desc.SetOutput("Out", {"mul_out"});
-  int num_flatten_dims = 3;
-  desc.SetAttr("x_num_col_dims", num_flatten_dims);
+  // int num_flatten_dims = 3;
```
Comment (Superjomn) [inline paddle/fluid/inference/anakin/convert/test_fc_op.cc]: clean these

File: paddle/fluid/inference/anakin/convert/test_fc_op.cc [hunk 2]
```diff
@@ -22,26 +21,24 @@ namespace inference {
...
+  // desc.SetAttr("x_num_col_dims", num_flatten_dims);
```
Comment (Superjomn) [inline paddle/fluid/inference/anakin/convert/test_fc_op.cc]: clean these

File: paddle/fluid/inference/api/paddle_api.h [hunk 3]
```diff
@@ -177,6 +177,8 @@ class ZeroCopyTensor {
     device_ = device;
   }
 
+  PaddleDType type();
```
Comment (Superjomn) [inline paddle/fluid/inference/api/paddle_api.h]:
```
 PaddleDType type() const;
```

File: paddle/fluid/operators/anakin/anakin_engine_op.h [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                             const platform::Place &dev_place) const {
+    if (anakin_engine_ == nullptr) {
+      anakin_engine_ =
+          inference::Singleton<inference::anakin::AnakinEngineManager>::Global()
+              .Get(engine_key_);
+    }
+    // BUG here, detect that the tensor data pointer here will change sometime.
+    // Will fix it later.
+    /*
+    // For share with the tensor from fluid, We do the net init in the first net
```
Comment (Superjomn) [inline paddle/fluid/operators/anakin/anakin_engine_op.h]: clean these

recover these if valid latter

File: paddle/fluid/inference/anakin/convert/test_fc_op.cc (hunk 1 above)
Comment (NHZlX) [reply paddle/fluid/inference/anakin/convert/test_fc_op.cc]: Done

File: paddle/fluid/inference/anakin/convert/test_fc_op.cc (hunk 2 above)
Comment (NHZlX) [reply paddle/fluid/inference/anakin/convert/test_fc_op.cc]: done

File: paddle/fluid/inference/api/paddle_api.h (hunk 3 above)
Comment (NHZlX) [reply paddle/fluid/inference/api/paddle_api.h]: Done

File: paddle/fluid/operators/anakin/anakin_engine_op.h (hunk 4 above)
Comment (NHZlX) [reply paddle/fluid/operators/anakin/anakin_engine_op.h]: Done

**Review by Superjomn — APPROVED** [review]
LGTM

**Pull request merged**